pkg/vuln/nvdapi/detector.go

```diff
@@ -0,0 +1,434 @@
+// rate limiter.
+func (s *Detector) VulnerabilitiesForPackages(ctx context.Context, packages ...string) (map[string][]vuln.Match, error) {
+	matchesByPackage := make(map[string][]vuln.Match)
+	matchesByPackageMutex := new(sync.Mutex) // avoid map concurrency issues
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	for _, pkg := range packages {
+		pkg := pkg // https://go.dev/doc/faq#closures_and_goroutines
+
+		g.Go(func() error {
+			matches, err := s.vulnerabilitiesForPackage(ctx, pkg)
+			if err != nil {
+				return err
+			}
+
+			if count := len(matches); count >= 1 {
+				log.Printf("🤔 %s: potential CVE matches: %d", pkg, count)
+			} else {
+				log.Printf("😅 %s: no CVE matches returned", pkg)
+			}
+
+			matchesByPackageMutex.Lock()
+			matchesByPackage[pkg] = matches
+			matchesByPackageMutex.Unlock()
+
+			return nil
+		})
+	}
+
+	err := g.Wait()
+	if err != nil {
+		return nil, err
+	}
+
+	return matchesByPackage, nil
+}
+
+func (s *Detector) vulnerabilitiesForPackage(ctx context.Context, name string) ([]vuln.Match, error) {
+	requestCPE := s.getCPE(name)
+
+	var result []vuln.Match
+
+	cves, err := s.doSearch(ctx, requestCPE)
+	if err != nil {
+		return nil, err
+	}
+
+	//nolint:gocritic // (rangeValCopy) for readability
+	for _, cve := range cves {
+		cve := cve
+		match, err := determineVulnMatch(&cve, name, requestCPE)
+		if err != nil {
+			return nil, err
+		}
+		if match == nil {
+			continue
+		}
+		result = append(result, *match)
+	}
+
+	return result, nil
+}
+
+// determineVulnMatch looks at the Cve response data to determine whether
+// there's a true match to the package. It returns an error if it can't produce
+// a vuln.Match. It also outputs a nil vuln.Match in non-error scenarios if the
+// Cve response data doesn't apply to the package.
+func determineVulnMatch(cve *Cve, packageName, requestCPE string) (*vuln.Match, error) {
+	// TODO: Should we only use the _first_ configuration, like secfixes-tracker does?
+
+	// TODO: Consider dropping CVE matches for really old CVEs (because the
+	//  data tends to be lower quality).
+
+	if cve == nil {
+		return nil, errors.New("input CVE was nil")
+	}
+
+	for _, configuration := range cve.Configurations {
+		if configuration.Operator == "AND" {
+			// For now, let's ignore AND-ed nodes. They usually mean that something is
+			// running on a specific platform, which usually means it's not the software
+			// we're looking for. E.g. https://nvd.nist.gov/vuln/detail/CVE-2009-3654.
+
+			continue
+		}
+
+		for _, node := range configuration.Nodes {
+			if node.Negate {
+				// TODO: Decide what to do with this case and why.
+
+				continue
+			}
+
+			for _, cpeMatch := range node.CpeMatch {
+				doCPEsMatch, err := cpeStringsMatch(requestCPE, cpeMatch.Criteria)
+				if err != nil {
+					return nil, err
+				}
+				if !doCPEsMatch {
+					continue
+				}
+
+				if !cpeMatch.Vulnerable {
+					continue
+				}
```

Comment:
```suggestion
				if !cpeMatch.Vulnerable {
					continue
				}
				
				doCPEsMatch, err := cpeStringsMatch(requestCPE, cpeMatch.Criteria)
				if err != nil {
					return nil, err
				}
				if !doCPEsMatch {
					continue
				}
```

Then if `!cpeMatch.Vulnerable` you never even have to do `cpeStringsMatch`. It's a tiny thing, if you think it's easier to read the other way that's fine, it just bails as quickly and early-ily as possible.

Reply:
Ah I got it! I like it, I'll do this. I had thought you meant rearranging the lines would mean some lines weren't needed.